### ISSUE TITLE
Slow performance

### ISSUE THREAD

**Issue by vmalloc** [fcsonline/drill#9]
On a 2018 Macbook Pro I'm unable to get drill to pass the ~140reqs/sec mark. `apib`, for comparison, easily goes to ~24k reqs/sec  on the same setup

**Comment by fcsonline** [comment]
Drill executes the test plan sequentially because you can have data dependencies between requests. If you are interested in concurrency, you can increase the number of threads.

**Comment by fcsonline** [comment]
I recorded the [this demo](https://asciinema.org/a/164615?autoplay=1) in a Linux container with a 2014 Dell XPS, reaching these numbers:

```
Concurrency Level 20                                                                                                                                                                                               
Time taken for tests 14.2 seconds                                                                                                                                                                                  
Total requests 10200                                                                                                                                                                                               
Successful requests 8400                                                                                                                                                                                           
Failed requests 1800                                                                                                                                                                                               
Requests per second 716.856 [#/sec]                                                                                                                                                                                
Time per request 1ms                                                                                                                                                                                               
Sample standard deviation 0ms
```
This demo has dependencies between requests. I have to say that the target is against localhost server to avoid network latency.

As I said, each iteration is waiting to complete each request, though it has only one request. I'm not sure if `apib` is just running as much requests as possible in parallel.

**Comment by vmalloc** [comment]
So what is the pattern I should be using to just test how many parallel requests to `/` my app can take?

**Comment by fcsonline** [comment]
I have been thinking about this use case and this project is not ready to request with the maximum throughput possible. We need to work with the asynchronous version of Hyper with Tokio promises to request them independently from the test plan.

**Comment by tzaeru** [comment]
At least on Windows, I don't seem to get any difference in throughput whether the amount of threads is 1 or 10. All requests appear to be made sequentially.

**Comment by fcsonline** [comment]
All requests defined in the test plan are executed sequentially per each thread, because you can have dependencies between them.

**Comment by tzaeru** [comment]
Right, but I don't seem to get any chance in requests-per-second when enabling more threads. So the threads wait for each other to finish as well?

**Comment by tzaeru** [comment]
Great, thanks for the fix! Now I get +300 requests/second, against a computer on another network. Actually managed to crash the office router.. Should probably have ran that on a development server!

**Comment by fcsonline** [comment]
Cool! 💪  I had a similar issue crashing one server with `Connection refused` errors. 😅 This is a good trace that the throughput has increased.  

To confirm that `drill` runs in parallel I used this test plan and this server:

- https://github.com/fcsonline/drill/blob/master/example/threads.yml
- https://github.com/fcsonline/drill/blob/master/example/responses/server.py

On the other hand, I added two new available interpolations that helped me to spot the problem:

- `{{ thread }}`: Thread identifier
- `{{ iteration }}`: Current iteration in the thread

**Comment by fcsonline** [comment]
I just published release `0.4.0` with those bug fixes. With it, I reach +500 requests/s.

Can I close this issue?

**Comment by tzaeru** [comment]
At least from my behalf, yeah, closable.

I actually manage more requests per second now with Drill than with a naive (aka default configs) Artillery setup. 3 times more requests, in fact. This is very good for our use case where we need test against a lot of requests per second.
